src/main/java/bio/terra/common/tracing/TracingHook.java

```diff
@@ -0,0 +1,205 @@
+  }
+
+  /**
+   * Creates a Flight Span for the current run of the Flight. Note that the same Flight may be run
+   * multiple times by Stairway.
+   */
+  @Override
+  public HookAction startFlight(FlightContext context) {
+    SpanContext submissionContext = getOrSubmissionContext(context);
+    // Start the Flight Span. We must remember to close the Flight Span at the end of the Flight's
+    // current run.
+    Span flightSpan =
+        Tracing.getTracer()
+            .spanBuilder(FLIGHT_NAME_PREFIX + context.getFlightClassName())
+            .startSpan();
+    flightSpan.addLink(Link.fromSpanContext(submissionContext, Link.Type.PARENT_LINKED_SPAN));
+    flightSpan.putAttribute("stairway/flightId", AttributeValue.stringAttributeValue(context.getFlightId()));
+    flightSpan.putAttribute(
+        "stairway/flightClass", AttributeValue.stringAttributeValue(context.getFlightClassName()));
+    flightSpans.put(context.getFlightId(), flightSpan);
+    return HookAction.CONTINUE;
+  }
+
+  /** Ends the Flight Span for the current run of the Flight. */
+  @Override
+  public HookAction endFlight(FlightContext context) {
+    Span flightSpan = flightSpans.remove(context.getFlightId());
+    // End the Flight Span for the current run.
+    flightSpan.end();
+    return HookAction.CONTINUE;
+  }
+
+  @Override
+  public Optional<StepHook> stepFactory(FlightContext context) {
+    return Optional.of(new TraceStepHook());
+  }
+
+  /** A {@link StepHook} for creating Spans for each Step execution. */
+  private class TraceStepHook implements StepHook {
+    private Span stepSpan = null;
+    private Scope stepScope = null;
+
+    @Override
+    public HookAction startStep(FlightContext flightContext) {
+      Span flightSpan = flightSpans.get(flightContext.getFlightId());
+
+      // Start the Step Span. We must remember to close the Span at the end of the Step.
+      stepSpan =
+          Tracing.getTracer()
+              .spanBuilderWithExplicitParent(
+                  STEP_NAME_PREFIX + flightContext.getStepClassName(), flightSpan)
+              .startSpan();
+      stepSpan.putAttribute(
+          "stairway/flightId", AttributeValue.stringAttributeValue(flightContext.getFlightId()));
+      stepSpan.putAttribute(
+          "stairway/flightClass", AttributeValue.stringAttributeValue(flightContext.getFlightClassName()));
+      stepSpan.putAttribute(
+          "stairway/stepClass", AttributeValue.stringAttributeValue(flightContext.getStepClassName()));
+      stepSpan.putAttribute("stairway/direction", AttributeValue.stringAttributeValue(flightContext.getDirection().toString()));
+      // Start the Scope of the Step Span's execution. We must remember to close the Scope at the
+      // end of the Step.
+      stepScope = tracer.withSpan(stepSpan);
+      return HookAction.CONTINUE;
+    }
+
+    @Override
+    public HookAction endStep(FlightContext flightContext) {
+      stepScope.close();
```

Comment:
I'm not hip to scopes. I notice you don't use one for flights. How come?

Reply:
The scope changes what the "current span" is, e.g. what `Tracing.getTracer().getCurrentSpan()` returns (implemented with ThreadLocals I believe). It's important for the step execution to run within a scope so that each step can add to the current span and get the right results. As Stairway doesn't do any span additions by itself, changing what the current span is for the flight isn't necessary, but also isn't harmful (as long as `startFlight`/`endFlight` are called on the same thread as `startStep`/`endStep`. As is, I'd have to add a pair or second map of `flightScopes` or add my own value class to store the Flight Scope in addition to the Flight Span, so it didn't seem worth it. Less work if we had a `FlightHook`.
Tried to add some more explanation to the comments.